src/watch.rs

```diff
@@ -26,98 +27,116 @@ where
 {
     println!("Watching for changes...");
 
-    let rx_changes = trigger_on_change(&mut watch_sources, &mut watch_code)?;
+    let mut watch_sources_old: HashSet<_> = watch_sources.collect();
+    let mut watch_code_old: HashSet<_> = watch_code.collect();
+
+    trigger_on_change(
+        watch_sources_old.clone(),
+        watch_code_old.clone(),
+        |change, sw, cw| {
+            if allow_reverse || change == ChangeType::Sources {
+                println!(
+                    "{} changed. Re-building...",
+                    if change == ChangeType::Sources {
+                        "Sources"
+                    } else {
+                        "Code"
+                    }
+                );
+
+                let curr_dir = env::current_dir()?;
+                let (config, mut watch_sources_new, watch_code_new, _has_reverse) =
+                    cmd::run_with_args(&args, Some(change == ChangeType::Code))?;
+                env::set_current_dir(&curr_dir)?;
+
+                watch_sources_new.insert(config);
 
-    for change in rx_changes {
-        if allow_reverse || change == ChangeType::Sources {
-            println!(
-                "{} changed. Re-building...",
-                if change == ChangeType::Sources {
-                    "Sources"
-                } else {
-                    "Code"
+                for path in &watch_sources_old {
```

Comment:
I think if you go through the trouble of updating the existing watchers (instead of just creating a completely new set of watchers/threads), you should extend this to only remove old paths which are not also part of the new paths and to add only those new paths which where not already part of the old paths.

Reply:
This can indeed be simplified further...